src/compiler/transformers/jsx.ts

```diff
@@ -24,7 +26,458 @@ namespace ts {
         }
 
         function visitorWorker(node: Node): Node {
-            return node;
+            switch (node.kind) {
+                case SyntaxKind.JsxElement:
+                    return visitJsxElement(<JsxElement>node);
+
+                case SyntaxKind.JsxSelfClosingElement:
+                    return visitJsxSelfClosingElement(<JsxSelfClosingElement>node);
+            }
+
+            Debug.fail("Unexpected node kind.");
+        }
+
+        function transformJsxChildToExpression(node: JsxChild): Expression {
+            switch (node.kind) {
+                case SyntaxKind.JsxText:
+                    return visitJsxText(<JsxText>node);
+
+                case SyntaxKind.JsxExpression:
+                    return visitJsxExpression(<JsxExpression>node);
+
+                case SyntaxKind.JsxElement:
+                    return visitJsxElement(<JsxElement>node);
+
+                case SyntaxKind.JsxSelfClosingElement:
+                    return visitJsxSelfClosingElement(<JsxSelfClosingElement>node);
+            }
+
+            Debug.fail("Unexpected node kind.");
+        }
+
+        function visitJsxElement(node: JsxElement) {
+            return visitJsxOpeningLikeElement(node.openingElement, node.children);
+        }
+
+        function visitJsxSelfClosingElement(node: JsxSelfClosingElement) {
+            return visitJsxOpeningLikeElement(node, /*children*/ undefined);
+        }
+
+        function visitJsxOpeningLikeElement(node: JsxOpeningLikeElement, children: JsxChild[]) {
+            // We must the node onto the node stack if it is not already at the top.

```

Comment:
I'll delete this. There is no "node stack" in this implementation of transforms.